### PR TITLE
handle player quitting

### PR DESCRIPTION
**Pull request by josephmaxwellhyland**

### CODE REVIEW

File: back-end/db/player.ts [hunk 1]
```diff
@@ -142,7 +142,7 @@ export default {
   },
 
   getPlayersSortedByGuessSpeed: function(guessingPlayers: IPlayer[], quizQIndex: number): IPlayer[] {
-    const guessTime = p => p.quizGuesses[quizQIndex].timestamp;
+    const guessTime = p => p.quizGuesses[quizQIndex] ? p.quizGuesses[quizQIndex].timestamp : 0;
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:145]: handle potential players who did not guess

File: back-end/handlers/hostHelpers.ts [hunk 2]
```diff
@@ -159,7 +160,8 @@ const handleTiebreakerAnswers = async (allPlayers: IPlayer[], quizQIndex, correc
   let bonusPlayer: IPlayer = bottomPlayersBySpeed[0];
   for (let i = 1; i < bottomPlayersBySpeed.length; i++) {
     const currentPlayer = bottomPlayersBySpeed[i];
-    if (currentPlayer.quizGuesses[quizQIndex].guess === correctQIndex) {
+    const currentGuess = currentPlayer.quizGuesses[quizQIndex];
+    if (currentGuess && currentGuess.guess === correctQIndex) {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:164]: handle potential players who did not guess

File: back-end/handlers/hostHelpers.ts [hunk 3]
```diff
@@ -211,11 +216,11 @@ const hostShowAnswer = async (gameId: number, io: Server): Promise<void> => {
     playerDb.updatePlayerState(subjectPlayer.id, newSubjectPlayerState, io, {});
   }
 
-  for (let i = 0; i < guessingPlayers.length; i++) {
-    const currentPlayer = guessingPlayers[i];
-    const currentPlayerCurrentGuess = currentPlayer.quizGuesses[currentQuestionIndex].guess;
-    const playerCorrect = currentPlayerCurrentGuess === correctGuess;
-    const currentPlayerNewState = playerCorrect ? PlayerStates.SeeingAnswerCorrect : PlayerStates.SeeingAnswerIncorrect;
+  for (let i = 0; i < nonSubjectPlayers.length; i++) {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:219]: make sure everyone is notified one way or another

File: back-end/handlers/hostHelpers.ts [hunk 4]
```diff
@@ -263,12 +268,24 @@ const onHostViewUpdate = async(gameId, io: Server) => {
...
+    } else if (gameData.gameState.state === GameStates.Questionnaire) {
       await hostStartQuiz(gameId, io);
     }
   } catch (e) {
     io.to(gameData.hostSocketId).emit("onHostViewUpdate-error", e);
   }
 }
 
-export default { hostStartQuiz, hostPreAnswer, onHostViewUpdate,  hostNextQuestionOrLeaderboard, hostSkipTimer, hostShowIntLeaderboard, hostStartQuizTimer, hostPreLeaderBoard };
+const handlePlayerQuit = async function(player: IPlayer, game: IGame, io: Server) {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:279]: handle player-quit message from a player

File: back-end/handlers/playerHandler.ts [hunk 5]
```diff
@@ -141,11 +141,29 @@ export default (io: Server, socket: Socket) => {
       await hostHelpers.onHostViewUpdate(gameId, io);
     } catch (e) {
       console.error("Failed to kick player: " + e);
-    }}
+    }
+  };
+
+  const onPlayerQuit = async () => {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/playerHandler.ts:147]: respond to player quitting

File: front-end/src/host/HostIntermediaryLeaderBoard.tsx [hunk 6]
```diff
@@ -58,11 +58,18 @@ export default function HostIntLeaderBoard(props: IntLeaderboardProps) {
...
-  );
+  let randomMessage;
+  if (playerScores.length > 1) {
+    randomMessage = randomMessageGenerator(
+      playerScores[0],
+      playerScores[1],
+      playerScores[playerScores.length - 1]
+    );
+  } else if (playerScores.length === 1) {
+    randomMessage = "Congrats on being the only player here, " + playerScores[0].name;
```
Comment (josephmaxwellhyland) [inline front-end/src/host/HostIntermediaryLeaderBoard.tsx:69]: handle potential games with fewer than 2 people

File: front-end/src/player/PlayerNewRanking.tsx [hunk 7]
```diff
@@ -69,7 +69,7 @@ export default function PlayerNewRanking(props: RankingProps) {
 
   return (
     <>
-      <div>{getPlayerRankDisplay()}</div>
+      <div>{playerScores.length > 1 ? getPlayerRankDisplay() : "What are you doing here?"}</div>
```
Comment (josephmaxwellhyland) [inline front-end/src/player/PlayerNewRanking.tsx:72]: what if there's only one player

**Review by hylandtechoutreach — APPROVED** [review]

**Pull request merged**